### PR TITLE
fix: deduplicate mget response

### PR DESCRIPTION
**Pull request by romange**
In case of duplicate mget keys, skips fetching the same key twice.
We add transactional flag DUPLICATE_KEYS_HINT that tells us for free
whether shard has duplicate keys.
The optimization is straighforward - we just copy the response for the original key,
since the response is a shallow object, we potentially save lots of memory with this
deduplication.

### CODE REVIEW

File: src/server/transaction.cc [hunk 1]
```diff
@@ -1104,10 +1104,13 @@ bool Transaction::ScheduleInShard(EngineShard* shard, bool execute_optimistic) {
 
     // We need to acquire the fp locks because the executing callback
     // within RunCallback below might preempt.
-    bool keys_unlocked = GetDbSlice(shard->shard_id()).Acquire(mode, lock_args);
+    auto [keys_unlocked, has_duplicates] = GetDbSlice(shard->shard_id()).Acquire(mode, lock_args);
```
Comment (adiholden) [inline src/server/transaction.cc:1107]: when running dragonfly with lock on hash tag will this always be true for multi keys?
if so do you see any downside?

Comment (adiholden) [reply src/server/transaction.cc:1107]: I think we are missing some tests that check the response for mget for the same key
lets add this test before merging this pr

Comment (romange) [reply src/server/transaction.cc:1107]: good catch, i did not think of it. yes, there is a downside

Comment (romange) [reply src/server/transaction.cc:1107]: I think I am gonna drop the transaction optimization for now, too bad it does not work.

**Comment by adiholden** [comment]
@romange please add unit test for checking the output of mget with same key

**Comment by romange** [comment]
@adiholden  MGetCachingModeBug2465 that you added already covers it.
Also SetWithHashtagsWithEmulatedCluster covers the hashtags scenario.

**Comment by adiholden** [comment]
> MGetCachingModeBug2465

As I wrote above we do not check the mget response, we just run the mget command and check the stats
you can change this test to check the mget response

File: src/server/string_family.cc [hunk 2]
```diff
@@ -536,37 +537,59 @@ struct MGetResponse {
...
-  absl::InlinedVector<DbSlice::ConstIterator, 32> iters(keys.Size());
+  struct Item {
+    DbSlice::ConstIterator it;
+    int source_index = -1;  // in case of duplicate keys, points to the first occurrence.
+  };
+
+  absl::InlinedVector<Item, 32> items(keys.Size());
+  static thread_local absl::flat_hash_map<string_view, unsigned> key_index;
+
+  FiberAtomicGuard guard;
```
Comment (adiholden) [inline src/server/string_family.cc:556]: I dont think this is not preemptive
If you run in cache mode and do bumpups for example you will need to serialize the bucket

Comment (romange) [reply src/server/string_family.cc:556]: then we can not use thread_local.

Comment (romange) [reply src/server/string_family.cc:556]: you good you.

File: src/server/string_family.cc [hunk 3]
```diff
@@ -596,6 +623,7 @@ MGetResponse OpMGet(util::fb2::BlockingCounter wait_bc, uint8_t fetch_mask, cons
       }
     }
   }
+  key_index.clear();
```
Comment (adiholden) [inline src/server/string_family.cc:626]: ok this is redundant now

File: src/server/string_family.cc [hunk 4]
```diff
@@ -536,37 +537,63 @@ struct MGetResponse {
...
+    DbSlice::ConstIterator it;
+    int source_index = -1;  // in case of duplicate keys, points to the first occurrence.
+  };
+
+  absl::InlinedVector<Item, 32> items(keys.Size());
+
+  // We can not make it thread-local because we may preempt during the Find loop due to
+  // serialization with the bumpup calls.
+
+  // TODO: consider separating BumpUps from finds because it becomes too complicated
```
Comment (adiholden) [inline src/server/string_family.cc:558]: I just think that we can not assume that find can not preempt
weather its find read only or find mutable

**Pull request merged**